### DISCUSSION TITLE
Band structure prediction

### DISCUSSION THREAD

**Discussion by keysongkang** [mzjb/DeepH-pack#50] [Q&A] [answered]
Currently, I'm in the final stage of "inference with the model," and I have a few questions that have arisen during this step.

1. When setting the value for 'band.json,' what should I input for the fermi_level? Should it be the ground state Fermi level?

2. In this case, I am testing the silicon structure containing 64 atoms with total of 1152 orbitals. The training data was obtained from an MD simulation at 300 K. The model then predicts the band structure of a configuration from the testing data. For the fermi_level, I simply used 0.0, and the resulting plot is as follows: black lines represent data from DFT, while red lines represent data from DeepH. The left side of the plot shows the entire energy range, while the right side presents a zoomed-in view around 0.
![image](https://github.com/mzjb/DeepH-pack/assets/130345610/5df1f197-8c20-4374-ac3d-442108492afc)

3. One peculiar observation is that the number of bands from DFT and DeepH differs. I employed the "dense_jl" eigen_solver to avoid specifying "max_iter" and "num_band." The number of bands obtained from DFT is 640, whereas DeepH yields 1152 bands.

4. I also inspected the matrix elements of the Hamiltonians from DFT SCF calculation and DeepH.
![image](https://github.com/mzjb/DeepH-pack/assets/130345610/14fc128f-232b-4a77-af75-f27d2a8c6245)

At this point, I am unsure about what steps to take next. I feel a bit lost and unsure of what to do. I would greatly appreciate any comments or suggestions you may have. Thank you for your assistance!

**Comment by mzjb** [comment]
1. You can obtain the Fermi energy in DeepH by diagonalizing all valence bands or using the "selected inversion" algorithm. See https://github.com/mzjb/DeepH-pack/issues/20#issuecomment-1339399004
2. You should use the proper Fermi level to plot the band structure in order to compare the bands. Also, please note that DeepH uses energy units in eV, so make sure you compare both sides using the same unit of energy.
3. "dense_jl" calculates all the bands, which is equal to the number of orbitals. DFT programs may not calculate all the bands, for example, the default value might be several times more than the number of valence electrons.
4. It appears that the diagonal elements of the matrix elements are roughly the same ([0, 0] matrix element), indicating there is no significant overall shift in the Fermi energy. This means that you can directly use the Fermi energy calculated from DFT for the band calculation and plotting in DeepH, enabling you to make comparisons.

**Reply by keysongkang** [reply]
Thank you for your comment. I appreciate your input. I have carefully reviewed the units and adjusted the Fermi level accordingly. However, I must admit that the improvement in the results was not significant.

![image](https://github.com/mzjb/DeepH-pack/assets/130345610/d7f14551-7821-4521-bf92-48c4fd78660e)

Considering that the predicted Hamiltonian is not drastically different from the actual Hamiltonian (in my opinion), I find myself perplexed as to why the predictions are diverging so noticeably.

**Reply by mzjb** [reply]
You can start by adjusting the y-axis of the band structure to a range of a few eV around the Fermi level and see if the band shape computed by DeepH appears reasonable. If the band shape is not quite reasonable, you might need to check if there are any errors in the overlap matrix used for computing the band structure.

**Reply by keysongkang** [reply]
Thank you. I don't think shifting bands works.
Hmm... in that case, the overlap matrix could be a potential factor.

Do you know if the overlap matrix undergoes any changes after the SCF steps, in comparison to the overlap of atomic basis orbitals? I'm thinking about examining the overlap matrix from one of the training data to assess how accurately the model predicts the band structure of the trained configuration.

**Reply by keysongkang** [reply]
it seems that the overlap matrix element remains unchanged when comparing the one derived from the trained data with the one extracted without SCF steps.
![image](https://github.com/mzjb/DeepH-pack/assets/130345610/040bec5a-b113-40ff-a844-4a0f8e53a36d)

However, the trained model fails to accurately predict the band structure of the training data.
![image](https://github.com/mzjb/DeepH-pack/assets/130345610/a9c77b3d-7a50-41f8-8e88-91a9485a47c5)

**Reply by keysongkang** [reply]
I also tested "sparse_jl" and "dense_jl" options but the result doesn't change

**Reply by keysongkang** [reply]
When I compare the training data in zenodo, "info.json" is somewhat different from what I have.

MoS2 case
`{
    "nsites": 75,
    "fermi_level": -4.966581824433852,
    "isorthogonal": false,
    "isspinful": false,
    "norbits": 1125
}`

My case 
`{
    "isspinful": false
}
`

Would that matter?

**Reply by mzjb** [reply, marked as answer]
This difference is not a problem.

**Reply by keysongkang** [reply]
Thank you for your comment. I appreciate your input and I'm doing my best to investigate all possible reasons for the issue. However, I haven't come across any clear clues or solutions yet. As part of my troubleshooting process, I'm planning to run the tutorial to better understand its functionality and determine if it can help resolve the problem.

If you don't mind, would you be able to take a look at my input files? It would be a great help for me to find the solution.
https://www.transfernow.net/dl/202307067mDGZ4Ct

**Reply by keysongkang** [reply]
I could reproduce the tutorial example. Then, it is a problem that happens during the FHI-aims process.
